Performance benchmark + visualization

HereThereBeDragons opened the pull request:
currently user-defined parameters must be changed in `start_... .py`

if everything looks good we should go should add default settings and let the rest be overwritten by command line arguments

HereThereBeDragons committed:
wip python performance benchmark (used for chep23)

HereThereBeDragons committed:
some fixes + more comments

HereThereBeDragons committed:
better printing

HereThereBeDragons committed:
add more comments

HereThereBeDragons committed:
working version for visualization but still needs more documentation/comments

HereThereBeDragons committed:
comments and cleanup/nicer plots for visualization

HereThereBeDragons committed:
remove unnecessary commands and colors

HereThereBeDragons committed:
everyone likes progress bars

HereThereBeDragons committed:
better variable content

HereThereBeDragons committed:
more comments

HereThereBeDragons committed:
remove whitespace and unnecessary output, better tutorial names

HereThereBeDragons committed:
wip

HereThereBeDragons committed:
start_visualization now works with YAML. still a bit cleanup missing

HereThereBeDragons committed:
visulizer = working version, missing is description/help + license

HereThereBeDragons committed:
WIP maybe working: start_benchmark use YAML config

HereThereBeDragons committed:
visualization: add ignore config checking for sections not being plotted/written

HereThereBeDragons committed:
add support for benchmarking non-cvmfs commands and to use the current (rpm) cvmfs installation (skip the install step)

HereThereBeDragons committed:
fix wrong indentation of exit() in benchmark_read_params.py

HereThereBeDragons committed:
bugfix: broken timeme function due to enabling autofs

HereThereBeDragons committed:
small fix and prettier output

vvolkl commented on test/performance-benchmark/util_benchmark/benchmark_cmds.py:
```diff
@@ -0,0 +1,117 @@
+################################################################################
+################################################################################
+
+#        ANYTHING BELOW HERE: DEPREACTED AND NOT USED ANYMORE
```
Please remove

HereThereBeDragons replied:
done

vvolkl commented on test/performance-benchmark/util_visualization/visualization_time.py:
```diff
@@ -0,0 +1,197 @@
+  "inode_cache.n_update": "Inode cache: #Updates",
+  "inode_cache.n_update_value": "Inode cache: #Value changes",
+  "inode_cache.sz_allocated": "Inode cache: #Allocated bytes",
+  "inode_cache.sz_size": "Inode cache: Total size", # TODO which size? bytes?
```
Yes, I believe this is given in bytes. Same below.

HereThereBeDragons replied:
done

vvolkl reviewed (commented):
Nice, thanks! I think this PR could benefit from one more round of cleanup, to remove stale/commented code.

We'll also want to move this in a `client` subfolder as we're adding server performance benchmark and more complicated setups, but it can be done in a future PR.

HereThereBeDragons committed:
change requests; move benchmark to client/; make writing of cvmfs internal affairs raw output optional

HereThereBeDragons commented:
also moved everything to subdir `performance-benchmark/client`

HereThereBeDragons committed:
create csv: add missing data column to output

HereThereBeDragons committed:
update gitignore, for benchmark: add send-thread-id which allows to send the thread id to the command executed; replace wait() with poll() for seeing if thread have finished executing the cmd to benchmark

HereThereBeDragons committed:
remove whitespace

HereThereBeDragons committed:
some cleanup in the benchmark scripts

HereThereBeDragons committed:
add more output to the logging file, add bash requirement to scripts,  some bug fixes

HereThereBeDragons committed:
commenting out to preload cache, and remove some other things

HereThereBeDragons committed:
add script: tensorflow for ubuntu22

HereThereBeDragons committed:
add support for benchmarking: run cvmfs unprivileged (only hot cache possible) and add option to preload cvmfs proxy cache (in avail_cmd)

HereThereBeDragons committed:
enforce unique keys for benchmark config reader (otherwise they would be silently overwritten)

HereThereBeDragons committed:
Visualization:
 1) Refactor of yaml config style to allow multiple plots of the same type decribed in the same config file.
 2) Refactor to allow for correct color schema if only unprivileged (= hot cache) data is visualized
 3) Extend labels for to visualize more parameters measured by "time"
 4) Bug fix where append_to_csv wrote out the wrong number of repetitions

HereThereBeDragons committed:
add example config to run and visualize tensorflow
